lib/OpenQA/WebAPI/Controller/API/V1/Job.pm

```diff
@@ -219,60 +220,174 @@ sub overview {
 
 =over 4
 
-=item create()
+=item _eval_param_grouping()
 
-Creates a job given a list of settings passed as parameters. TEST setting/parameter
-is mandatory and should be the name of the test.
+Removes job-specific parameters from the specified hash and returns a new hash
+containing the removed parameters (where the keys are job suffixes).
+
+So "{foo => 1, bar:jobA => 2, baz:jobB => 3}" will be turned into "{foo => 1}"
+returning "{jobA => {bar => 2}, jobB => {baz => 3}}".
 
 =back
 
 =cut
 
-sub create {
-    my $self = shift;
-    my $params = $self->req->params->to_hash;
-    my $is_clone_job = delete $params->{is_clone_job} // 0;
+sub _eval_param_grouping ($params) {
+    my %grouped_params;
+    for my $param_key (keys %$params) {
+        my $job_suffix_start = rindex($param_key, ':');
+        next unless $job_suffix_start >= 0;
+        my $job_suffix = substr $param_key, $job_suffix_start + 1;
+        my $setting_key = substr $param_key, 0, $job_suffix_start;
+        $grouped_params{$job_suffix}->{$setting_key} = delete $params->{$param_key};
+    }
+    return \%grouped_params;
+}
+
+=over 4
+
+=item _eval_dependency()
+
+Removes the specified $settings_keys from $job_settings populating $self->{_dependencies}
+with the dependency information from the removed hash entry value.
+
+=back
+
+=cut
 
+sub _eval_dependency ($self, $child_job_suffix, $job_settings, $setting_key, $type) {
+    return undef unless $child_job_suffix;
+    my $job_suffixes = delete $job_settings->{$setting_key};
+    return undef unless defined $job_suffixes;
+    my $deps = $self->{_dependencies};
+    for my $parent_job_suffix (split /\s*,\s*/, $job_suffixes) {
+        push @$deps,
+          {
+            child_job_id => $child_job_suffix,
+            parent_job_id => $parent_job_suffix,
+            dependency => $type,
+          };
+    }
+}
+
+=over 4
+
+=item _create_job()
+
+Creates a single job. The parameters $global_params and $job_specific_params will
+be merged ($job_specific_params takes precedence).
+
+=back
+
+=cut
+
+sub _create_job ($self, $global_params, $job_suffix = undef, $job_specific_params = {}) {
     # job_create expects upper case keys
-    my %up_params = map { uc $_ => $params->{$_} } keys %$params;
+    my %up_params = map { uc $_ => $global_params->{$_} } keys %$global_params;
+    $up_params{uc $_} = $job_specific_params->{$_} for keys %$job_specific_params;
+
     # restore URL encoded /
     my %params = map { $_ => $up_params{$_} =~ s@%2F@/@gr } keys %up_params;
-    if (!$params{TEST}) {
-        return $self->render(json => {error => 'TEST field mandatory'}, status => 400);
-    }
+    die "TEST field mandatory\n" unless $params{TEST};
```

Comment:
Maybe this just looks weird in the diff but just to double check: Are you changing the error rendering back into a die? Didn't you just recently have a PR doing the opposite?

Reply:
First, my other PR was about a whole different API. Second, this is still rendered as user-facing error because the controller code simply renders all `die` messages as such. (That isn't changed by my PR. A more fain-grained error handling would be nice but I consider it out of scope for now.)